test/mgh07.jl

```diff
@@ -0,0 +1,27 @@
+# MGH problem 7 - Helical valley function
+#
+#  Source:
+#  J. J. Moré, B. S. Garbow and K. E. Hillstrom
+#  Testing Unconstrained Optimization Software
+#  ACM Transactions on Mathematical Software, 7(1):17-41, 1981
+#
+# A. Montoison, Montreal, 05/2018.
+
+export mgh07
+
+"Helical valley function"
+function mgh07()
+
+  nls  = Model()
+  x0   = [-1.0,  0.0,  0.0]
+  @variable(nls, x[i=1:3], start = x0[i])
+
+  θ_aux(t) = (t > 0 ? 0.0 : 0.5)
+  JuMP.register(nls, :θ_aux, 1, θ_aux, autodiff=true)
+
+  @NLexpression(nls, F1, 10*(x[3] - 10*(atan(x[2]/x[1])/(2*π) + θ_aux(x[1]))))
+  @NLexpression(nls, F2, 10*(sqrt(x[1]^2 + x[2]^2) - 1.0))
+  @NLexpression(nls, F3, 1*x[3])
```

Comment:
And here, do we need to multiply by 1?

Reply:
In the file ```jump_model.jl```, the function ```MathProgNLSModel``` only accept ```Vector{JuMP.NonlinearExpression}```.

``` julia
function MathProgNLSModel(cmodel :: JuMP.Model,
                          F :: Vector{JuMP.NonlinearExpression}; kwargs...)
  # Hessian-Vector products don't work:
  # https://github.com/JuliaOpt/JuMP.jl/issues/1204
  @NLobjective(cmodel, Min, 0.5 * sum(Fi^2 for Fi in F))
  JuMP.setsolver(cmodel, ModelReader())
  cmodel.internalModelLoaded || JuMP.build(cmodel)
  ev = cmodel.internalModel.eval

  Fmodel = JuMP.Model()
  @NLobjective(Fmodel, Min, 0.0)
  Fmodel.nlpdata.user_operators = cmodel.nlpdata.user_operators
  @variable(Fmodel, x[1:MathProgBase.numvar(cmodel)])
  for Fi in F
    expr = ev.subexpressions_as_julia_expressions[Fi.index]
    replace!(expr, x)
    expr = :($expr == 0)
    JuMP.addNLconstraint(Fmodel, expr)
  end
  JuMP.setsolver(Fmodel, ModelReader())
  Fmodel.internalModelLoaded || JuMP.build(Fmodel)

  return MathProgNLSModel(Fmodel.internalModel,
                          cmodel.internalModel; kwargs...)
end
```
And JuMP want that we use ```@expression``` instead of ```@NLexpression``` when we have only a JuMP variable as expression. It's why we use the trick ```1*x[i]``` or ```x[i] + 0.0``` to allow it.

A solution to avoid it could be to accept both expression like that:
``` julia 
function MathProgNLSModel(cmodel :: JuMP.Model,
                          F :: Vector{JuMP.NonlinearExpression,JuMP.AffExpr}; kwargs...)
  # Hessian-Vector products don't work:
  # https://github.com/JuliaOpt/JuMP.jl/issues/1204
  @NLobjective(cmodel, Min, 0.5 * sum(Fi^2 for Fi in F))
  JuMP.setsolver(cmodel, ModelReader())
  cmodel.internalModelLoaded || JuMP.build(cmodel)
  ev = cmodel.internalModel.eval

  Fmodel = JuMP.Model()
  @NLobjective(Fmodel, Min, 0.0)
  Fmodel.nlpdata.user_operators = cmodel.nlpdata.user_operators
  @variable(Fmodel, x[1:MathProgBase.numvar(cmodel)])
  for Fi in F
  	if typeof(Fi) == JuMP.NonlinearExpression
    	expr = ev.subexpressions_as_julia_expressions[Fi.index]
    	replace!(expr, x)
    	expr = :($expr == 0)
    	JuMP.addNLconstraint(Fmodel, expr)
    else
    	c = GenericRangeConstraint(Fi,0.0,0.0)
    	JuMP.addconstraint(Fmodel, c)
    end
  end
  JuMP.setsolver(Fmodel, ModelReader())
  Fmodel.internalModelLoaded || JuMP.build(Fmodel)

  return MathProgNLSModel(Fmodel.internalModel,
                          cmodel.internalModel; kwargs...)
end
```
But I don't know the syntax to declare a vector that accept both types ```Vector{JuMP.NonlinearExpression,JuMP.AffExpr}```.

Reply:
I think something with [parametric types](https://docs.julialang.org/en/stable/manual/types/#Parametric-Types-1).